src/accountsdb/index.zig

```diff
@@ -85,47 +81,56 @@ pub const AccountIndex = struct {
                 break :blk .{ .disk = .{ .dma = disk_allocator, .ptr_allocator = allocator } };
             },
         };
-        _ = max_account_references; // TODO(fastload): use this
+
+        const reference_manager = try sig.utils.allocators.RecycleBuffer(AccountRef, .{}).create(.{
+            .memory_allocator = reference_allocator.get(),
+            .records_allocator = allocator,
+        });
 
         return .{
             .allocator = allocator,
-            .reference_allocator = reference_allocator,
             .pubkey_ref_map = try ShardedPubkeyRefMap.init(allocator, number_of_shards),
             .slot_reference_map = RwMux(SlotRefMap).init(SlotRefMap.init(allocator)),
+            .reference_allocator = reference_allocator,
+            .reference_manager = reference_manager,
         };
     }
 
-    pub fn deinit(self: *Self, free_memory: bool) void {
+    pub fn deinit(self: *Self) void {
         self.pubkey_ref_map.deinit();
 
         {
             const slot_reference_map, var slot_reference_map_lg = self.slot_reference_map.writeWithLock();
             defer slot_reference_map_lg.unlock();
-
-            if (free_memory) {
-                var iter = slot_reference_map.iterator();
-                while (iter.next()) |entry| {
-                    entry.value_ptr.deinit();
-                }
-            }
             slot_reference_map.deinit();
         }
 
+        self.reference_manager.deinit();
+        self.allocator.destroy(self.reference_manager);
         self.reference_allocator.deinit();
     }
 
-    pub fn putReferenceBlock(self: *Self, slot: Slot, references: std.ArrayList(AccountRef)) !void {
-        const slot_reference_map, var slot_reference_map_lg = self.slot_reference_map.writeWithLock();
-        defer slot_reference_map_lg.unlock();
-        try slot_reference_map.putNoClobber(slot, references);
-    }
+    pub fn deinitLoadingThread(self: *Self) void {
+        self.pubkey_ref_map.deinit();
+        {
+            const slot_reference_map, var slot_reference_map_lg = self.slot_reference_map.writeWithLock();
+            defer slot_reference_map_lg.unlock();
+            slot_reference_map.deinit();
+        }
 
-    pub fn freeReferenceBlock(self: *Self, slot: Slot) error{MemoryNotFound}!void {
-        const slot_reference_map, var slot_reference_map_lg = self.slot_reference_map.writeWithLock();
-        defer slot_reference_map_lg.unlock();
+        // // this is the main index's manager
+        // self.reference_manager.deinit();
+        // self.allocator.destroy(self.reference_manager);
+        // // dont free the references -- ownership is transferred to the main index
+        // for (self.reference_memory) |ref_block| {
+        //     self.reference_allocator.get().free(ref_block);
+        // }
+        // // the reference_allocator is the same as the main index's reference_allocator
+        // self.reference_allocator.deinit();
```

Comment:
Is this hanging?

Reply:
this is more of a comment which explains what we dont free and why relative to the normal deinit